doc/api/util.md

```diff
@@ -1097,6 +1097,80 @@ console.log(values, positionals);
 // Prints: [Object: null prototype] { foo: true, bar: 'b' } []
 ```
 
+Detailed parse information is available for adding custom behaviours by
+specifying `tokens: true` in the configuration. The returned tokens have
+properties describing:
+
+* all tokens
```

Comment:
Just a thought, I would be tempted to make tokens a class rather than a POJO with a `kind` property:

```js
class Token {}
class OptionTerminator extends Token {}
class PositionalToken extends Token {}
```

With the classes exposed on the `parseArgs` export:

```
const {Token, PositionalToken} = parseArgs
```

Then I think you could just document the tokens as classes, similar to `Buffer.blob`, nested under the top level `parseArgs` docs, similar to how the docs are laid out for the [JavaScript embedder API](https://nodejs.org/api/async_hooks.html#javascript-embedder-api).

Reply:
Even if these are made classes they should definitely expose a `kind` property - it's really awkward to work with tree nodes which lack such a property, which is why pretty much every parser in the world has them. (Similarly, the DOM has `nodeName` in addition to nodes being instances of distinct classes.)

Reply:
> Even if these are made classes they should definitely expose a kind property 

Sounds good to me. I think it might be worth the class based approach, so that people could opt to check instanceof, if they're so inclined?

Should we consider making kind a Symbol, or is a string preferable?

Reply:
Strings are traditional - they're a lot easier to work with, since you don't have to import them.

Reply:
> Just a thought, I would be tempted to make tokens a class rather than a POJO 

It did cross my mind as soon as I added the `kind` property. 😄 .

The examples of `Blob` and `AsyncResource` and `AsyncLocalStorage` are all objects that authors will create themselves, with methods and state, where I think a Class adds more value.

I'm not against classes as such, but not seeing benefits in this case.

Reply:
> I'm not against classes as such, but not seeing benefits in this case.

I would love to hear from another Node maintainer on this feedback specifically. On my mind was that it makes the inheritance more explicit, between the types of tokens, also it allows you to use `instanceof` to infer the type of token.

I don't know if this is worth the added complexity for not.

Reply:
@aduh95 any opinion on whether we should emit a POJO or a class? I could be convinced either way (wondering if there's a precedent we can lean on).

Reply:
> Using classes may help with documentation, but we could achieve the same with doc-only interfaces.

@aduh95 this sounds good to me.

You don't happen to know of any existing docs that follow this pattern that @shadowspawn could follow?

Reply:
I don't think there's a precedent, but I imagine it could look something like:


```markdown
### Interface: `PositionalToken`

<!-- YAML
added: REPLACEME
-->

* Extends: {Token}

#### `positionalToken.value`

 * {string} The value of the positional argument in args (i.e. `args[index]`).
```

Reply:
I had a look around in the Node.js docs, and was interested to see `fs.lstat` returns `fs.Stats` class (and not just a POJO): https://nodejs.org/dist/latest-v18.x/docs/api/fs.html#class-fsstats

That feels somewhat similar to tokens. An object that is probably not created or extended by an end-user, but is still a wrapped in an object for  completeness.

Reply:
My 2 cents is we should go with the first one and make it less verbose, the "existing" version can be a bit confusing I think. But feel free to disagree, none option is perfect anyway, and we can always edit the docs later if we find a better layout.

Reply:
I definitely find the existing one massively clearer; muddling things up with unnecessary TS/Java-ish terms and structure makes things more confusing.

Reply:
Mixed opinions straight away! Thanks both for feedback.

I will leave it as is, barring further feedback.

Reply:
I like that we could map the interface based approach directly to the TypeScript types we release on DefinitelyTyped.

But I don't feel super strongly, and would rather not block progress on the API (as @aduh95 says, we can edit the docs later).

Reapproving this PR 👍, @aduh95 are you happy as is?

Reply:
The compact POJO description is a bit subtle to read. How about an expanded version with all the properties listed?

## Proposed expanded

A returned token has two properties which are always defined,
and some other properties which vary depending on the `kind`:
  * `kind` {string} One of 'option', 'positional', or 'option-terminator'.
  * `index` {number} Index of element in `args` containing token. So the
    source argument for a token is `args[token.index]`.

An option token has additional parse details
for an option detected in the input `args`:
  * `kind` = `'option'`
  * `index` {number} Index of element in `args` containing token.
  * `name` {string} Long name of option.
  * `rawName` {string} How option used in args, like `-f` of `--foo`.
  * `value` {string | undefined} Option value specified in args.
    Undefined for boolean options.
  * `inlineValue` {boolean | undefined} Whether option value specified inline,
    like `--foo=bar`.

A positional token has just one additional property with the positional value:
  * `kind` = `'positional'`
  * `index` {number} Index of element in `args` containing token.
  * `value` {string} The value of the positional argument in args (i.e. `args[index]`).

An option-terminator token has only the base properties:
  * `kind` = `'option-terminator'`
  * `index` {number} Index of element in `args` containing token.

## Old compact

The returned tokens have properties describing:

* all tokens
  * `kind` {string} One of 'option', 'positional', or 'option-terminator'.
  * `index` {number} Index of element in `args` containing token. So the
    source argument for a token is `args[token.index]`.
* option tokens
  * `name` {string} Long name of option.
  * `rawName` {string} How option used in args, like `-f` of `--foo`.
  * `value` {string | undefined} Option value specified in args.
    Undefined for boolean options.
  * `inlineValue` {boolean | undefined} Whether option value specified inline,
    like `--foo=bar`.
* positional tokens
  * `value` {string} The value of the positional argument in args (i.e. `args[index]`).
* option-terminator token

(Duplicate post of: https://github.com/pkgjs/parseargs/pull/129#discussion_r922726703)